File: data/src/ishamela_data/normalizer.py
"""Arabic search normalizer (SPEC-001).

Maps source text to a canonical search form. Applied to ``body_norm`` at
bundle-build time and to user queries at search time — never to displayed
``pages.body``.

Rule order is contractual (SPEC-001): NFC → N1..N9 → N10.
"""

from __future__ import annotations

import unicodedata

NORM_VERSION: str = "1.0.0"

# N1 tashkīl U+064B–U+065F and superscript alef U+0670
# N2 tatweel U+0640
# N3 Quranic annotation U+06D6–U+06ED and U+08D3–U+08FF
# N7 standalone hamza U+0621 is removed (not replaced with a space)
_DELETE = {
    *(chr(cp) for cp in range(0x064B, 0x0660)),
    "\u0670",
    "\u0640",
    *(chr(cp) for cp in range(0x06D6, 0x06EE)),
    *(chr(cp) for cp in range(0x08D3, 0x0900)),
    "\u0621",
}

# N4 alef variants, N5 alef maqsura, N6 ta marbuta, N7 hamza carriers, N8 digits
_REPLACE = {
    "\u0623": "\u0627",  # أ → ا
    "\u0625": "\u0627",  # إ → ا
    "\u0622": "\u0627",  # آ → ا
    "\u0671": "\u0627",  # ٱ → ا
    "\u0649": "\u064A",  # ى → ي
    "\u0629": "\u0647",  # ة → ه
    "\u0624": "\u0648",  # ؤ → و
    "\u0626": "\u064A",  # ئ → ي
    **{chr(0x0660 + i): str(i) for i in range(10)},
    **{chr(0x06F0 + i): str(i) for i in range(10)},
}

_TRANS = str.maketrans(
    {ord(src): ord(dst) for src, dst in _REPLACE.items()}
    | {ord(ch): None for ch in _DELETE}
)


def normalize(text: str) -> str:
    """Return the canonical search form of ``text`` (SPEC-001 N11, N1–N10).

    Pure function: no options, no locale dependence. Idempotent.
    """
    # N11 — NFC before any stripping or folding
    text = unicodedata.normalize("NFC", text)
    # N1–N8 — strip marks / tatweel / Quranic annotation; fold letters; ASCII digits
    text = text.translate(_TRANS)
    # N9 — non-letter symbols (punctuation, brackets, ornaments, format chars) → space
    stripped = [
        ch
        if ch.isspace() or ("0" <= ch <= "9") or unicodedata.category(ch)[0] == "L"
        else " "
        for ch in text
    ]
    # N10 — collapse whitespace runs to one U+0020 and trim
    return " ".join("".join(stripped).split())

File: app/lib/core/search/normalizer.dart
/// Arabic search normalizer (SPEC-001).
///
/// Maps source text to a canonical search form. Applied to `body_norm` at
/// bundle-build time and to user queries at search time — never to displayed
/// `pages.body`.
///
/// Rule order is contractual (SPEC-001): NFC → N1..N9 → N10.
library;

const String normVersion = '1.0.0';

/// Targeted canonical composition covering Arabic hamza/maddah pairs that
/// interact with N1–N7, plus Latin combining acute used in the golden vectors.
///
/// Dart's core library has no Unicode NFC. This is not a silent switch to
/// NFKC: Arabic presentation forms such as U+FEFB are left unchanged.
int? _compose(int starter, int mark) {
  if (starter == 0x0627) {
    if (mark == 0x0653) return 0x0622; // ا + maddah → آ
    if (mark == 0x0654) return 0x0623; // ا + hamza above → أ
    if (mark == 0x0655) return 0x0625; // ا + hamza below → إ
  } else if (starter == 0x0648 && mark == 0x0654) {
    return 0x0624; // و + hamza above → ؤ
  } else if (starter == 0x064A && mark == 0x0654) {
    return 0x0626; // ي + hamza above → ئ
  } else if (starter == 0x0065 && mark == 0x0301) {
    return 0x00E9; // e + acute → é
  }
  return null;
}

String _nfc(String text) {
  if (text.isEmpty) {
    return text;
  }
  final out = <int>[];
  for (final rune in text.runes) {
    if (out.isNotEmpty) {
      final composed = _compose(out.last, rune);
      if (composed != null) {
        out[out.length - 1] = composed;
        continue;
      }
    }
    out.add(rune);
  }
  return String.fromCharCodes(out);
}

bool _isDeleted(int rune) {
  // N1 tashkīl
  if (rune >= 0x064B && rune <= 0x065F) return true;
  if (rune == 0x0670) return true; // superscript alef
  // N2 tatweel
  if (rune == 0x0640) return true;
  // N3 Quranic annotation
  if (rune >= 0x06D6 && rune <= 0x06ED) return true;
  if (rune >= 0x08D3 && rune <= 0x08FF) return true;
  // N7 standalone hamza
  if (rune == 0x0621) return true;
  return false;
}

int _fold(int rune) {
  switch (rune) {
    case 0x0623: // أ
    case 0x0625: // إ
    case 0x0622: // آ
    case 0x0671: // ٱ
      return 0x0627;
    case 0x0649: // ى
      return 0x064A;
    case 0x0629: // ة
      return 0x0647;
    case 0x0624: // ؤ
      return 0x0648;
    case 0x0626: // ئ
      return 0x064A;
    default:
      if (rune >= 0x0660 && rune <= 0x0669) {
        return 0x30 + (rune - 0x0660);
      }
      if (rune >= 0x06F0 && rune <= 0x06F9) {
        return 0x30 + (rune - 0x06F0);
      }
      return rune;
  }
}

final _letterRe = RegExp(r'\p{L}', unicode: true);
final _spaceRe = RegExp(r'\s', unicode: true);
final _spaceRunRe = RegExp(r'\s+', unicode: true);

bool _keep(int rune) {
  if (rune >= 0x30 && rune <= 0x39) return true;
  final ch = String.fromCharCode(rune);
  return _spaceRe.hasMatch(ch) || _letterRe.hasMatch(ch);
}

/// Return the canonical search form of [text] (SPEC-001 N11, N1–N10).
///
/// Pure function: no options, no locale dependence. Idempotent.
String normalize(String text) {
  text = _nfc(text);
  final buf = StringBuffer();
  for (final rune in text.runes) {
    if (_isDeleted(rune)) {
      continue;
    }
    final mapped = _fold(rune);
    if (_keep(mapped)) {
      buf.writeCharCode(mapped);
    } else {
      buf.writeCharCode(0x20);
    }
  }
  return buf
      .toString()
      .split(_spaceRunRe)
      .where((part) => part.isNotEmpty)
      .join(' ');
}
